fvm/state/transaction_state_test.go

```diff
@@ -523,3 +523,50 @@ func TestPauseAndResume(t *testing.T) {
 	require.NoError(t, err)
 	require.NotNil(t, val)
 }
+
+func TestFinalizeMainTransactionFailWithUnexpectedNestedTransactions(
+	t *testing.T,
+) {
+	txn := newTestTransactionState()
+
+	_, err := txn.BeginNestedTransaction()
+	require.NoError(t, err)
+
+	executionSnapshot, err := txn.FinalizeMainTransaction()
+	require.Error(t, err)
+	require.Nil(t, executionSnapshot)
+}
+
+func TestFinalizeMainTransaction(t *testing.T) {
+	txn := newTestTransactionState()
+
+	id1, err := txn.BeginNestedTransaction()
+	require.NoError(t, err)
+
+	registerId := flow.NewRegisterID("foo", "bar")
+
+	value, err := txn.Get(registerId)
+	require.NoError(t, err)
+	require.Nil(t, value)
+
+	_, err = txn.CommitNestedTransaction(id1)
+	require.NoError(t, err)
+
+	value, err = txn.Get(registerId)
+	require.NoError(t, err)
+	require.Nil(t, value)
+
```

Comment:
```suggestion
	// calling Get on another register to test that the register won't be added to the ReadSet,
	// because Commit is not called 
	value, err = txn.Get(flow.NewRegisterID("another_key", "which wont be added to the Readset"))
	require.NoError(t, err)
	require.Nil(t, value)
```

Reply:
this will actually modify the read set.

FinalizeMainTransaction is equivalent to CommitNestedTransaction in some sense.  Finalize prepares the main transaction for commit.  The reason it's called Finalize (instead of Commit) is because we may discard the execution snapshot (and re-execute the transaction) on read write conflicts